### ISSUE TITLE
The file '/etc/openvpn/vars' was not found

### ISSUE THREAD

**Issue by yell0w4x** [kylemanna/docker-openvpn#608]
Looks like kylemanna/openvpn:latest is broken. Also kylemanna/openvpn:2.3 works fine.

```
$ docker run -v $OVPN_DATA:/etc/openvpn --log-driver=none --rm -it kylemanna/openvpn ovpn_initpki

Easy-RSA error:

The file '/etc/openvpn/vars' was not found.
```

**Comment by valerioferretti92** [comment]
Same thing happened to me.

**Comment by anthonymobile** [comment]
for the dummies like me

```
docker run -v $OVPN_DATA:/etc/openvpn --log-driver=none --rm kylemanna/openvpn:2.3 ovpn_genconfig -u udp://vpn.yourdomain.com
docker run -v $OVPN_DATA:/etc/openvpn --log-driver=none --rm -it kylemanna/openvpn:2.3 ovpn_initpki

```

**Comment by yell0w4x** [comment]
> Also kylemanna/openvpn:2.3 works fine.
The issue is about kylemanna/openvpn:latest

**Comment by joeky888** [comment]
Workaround here https://github.com/kylemanna/docker-openvpn/issues/605#issuecomment-694836874

**Comment by RafayAK** [comment]
The same also happing to me, will try @joeky888 reference

**Comment by RafayAK** [comment]
using 2.3 works

**Comment by joshuacox** [comment]
hmmm, I can get the image running by using the 2.3 tag, but upon trying to create a  client:

```
Easy-RSA error:

EASYRSA_PKI does not exist (perhaps you need to run init-pki)?
Expected to find the EASYRSA_PKI at: /etc/openvpn/pki
Run easyrsa without commands for usage and command help.
failed to resize tty, using default size
/usr/local/bin/ovpn_getclient: line 16: /etc/openvpn/ovpn_env.sh: No such file or directory
```

**Comment by ebarault** [comment]
@kylemanna hi, i've been using the 2.4 tag from dockerhub for months now without any problem. I see now that you overwrote the 2.4 tag on dockerhub just 14 days ago. Are you aware of the problem described in this thread?

**Comment by kylemanna** [comment]
@ebarault I'm not aware of what the problem remaining is. I fixed a number of issues most importantly the CI tests running on Tavis-CI.  No issues are reported on the branch that docker tag points to:

https://travis-ci.org/github/kylemanna/docker-openvpn/builds/741565318

Does the problem remain? Can you show me details like docker image hash + error?

**Comment by kylemanna** [comment]
Pretty sure this is fixed, someone can let me know if it's not true.  CI builds pass and they'd fail on this if it was a problem.

**Comment by ZalgoNoise** [comment]
Hi Kyle, 

Thanks a lot for the hard work on this project.

I was finding a similar issue with a fork of your repo, but it seems to be an issue with EasyRSA, not any of your binaries. 

It's fixable by running the `easyrsa_export` binary, pointing to `/etc/openvpn/vars`. Then, `init_pki` will work fine:

```bash
bash /usr/local/bin/easyrsa_vars export > /etc/openvpn/vars
bash /usr/local/bin/ovpn_initpki
```

Hopefully this should point people having the same issue towards the right direction. I haven't inspected your code yet to see if this is referenced anywhere and if it's skipped somehow.

Cheers!

**Comment by kylemanna** [comment]
I understand this was an issue, but I've updated many thing to resolve things like this.  Can anyone show me the exact steps to reproduce this on any Docker Hub hosted image (must include hash of image)?

The Travis-CI builds work fine with new images and I'm convinced this problem was resolved sometime ago until someone can show me more.

**Comment by ebarault** [comment]
Hi @kylemanna 
on image **4fa457dfce43**
![image](https://user-images.githubusercontent.com/12000935/100637181-ddf2f280-3332-11eb-8727-874b26558066.png)
I had to use this trick to 
`touch /etc/openvpn/vars`
https://github.com/ebarault/ovenvpn-ansible-role/commit/e6b221329632720bb0c0fb9f9547d8cf8b56b4ad#diff-3d0ff1709ca48add100327bb2a468e6c508fb92a159c64c4f99ad1df89d9bddeR82

**Comment by kylemanna** [comment]
@ebarault I don't see anything like that.  Here's my console:

```
$ docker volume rm $OVPN_DATA
ovpn-data-example

$ docker pull kylemanna/openvpn:2.4

2.4: Pulling from kylemanna/openvpn
Digest: sha256:7586e816602b14bb00553816d09c6890891a6d7d62a446352cdba1c4c37456dc
Status: Image is up to date for kylemanna/openvpn:2.4
docker.io/kylemanna/openvpn:2.4

$ docker inspect kylemanna/openvpn:2.4 | jq '.[0].Id'
"sha256:4fa457dfce432adb384a04f268cd4b9b73b5d1f660e857a114a2505bbd71ef6e"

$ docker volume create --name $OVPN_DATA
ovpn-data-example

$ docker run -v $OVPN_DATA:/etc/openvpn --rm kylemanna/openvpn:2.4 ovpn_genconfig -u udp://VPN.SERVERNAME.COM
Processing PUSH Config: 'block-outside-dns'
Processing Route Config: '192.168.254.0/24'
Processing PUSH Config: 'dhcp-option DNS 8.8.8.8'
Processing PUSH Config: 'dhcp-option DNS 8.8.4.4'
Processing PUSH Config: 'comp-lzo no'
Successfully generated config
Cleaning up before Exit ...

$ docker run -v $OVPN_DATA:/etc/openvpn --rm -it kylemanna/openvpn:2.4 ovpn_initpki nopass
Note: using Easy-RSA configuration from: /etc/openvpn/vars

init-pki complete; you may now create a CA or requests.
Your newly created PKI dir is: /etc/openvpn/pki

Note: using Easy-RSA configuration from: /etc/openvpn/vars
Using SSL: openssl OpenSSL 1.1.1g  21 Apr 2020
Generating RSA private key, 2048 bit long modulus (2 primes)
...............................................................................+++++
.............................+++++
e is 65537 (0x010001)
You are about to be asked to enter information that will be incorporated
into your certificate request.
What you are about to enter is what is called a Distinguished Name or a DN.
There are quite a few fields but you can leave some blank
For some fields there will be a default value,
If you enter '.', the field will be left blank.
-----
Common Name (eg: your user, host, or server name) [Easy-RSA CA]:

CA creation complete and you may now import and sign cert requests.
Your new CA certificate file for publishing is at:
/etc/openvpn/pki/ca.crt

Note: using Easy-RSA configuration from: /etc/openvpn/vars
Using SSL: openssl OpenSSL 1.1.1g  21 Apr 2020
Generating DH parameters, 2048 bit long safe prime, generator 2
This is going to take a long time
.............................+...............+.............................................................................................................................................................................................+....................................................................................................+....................................+...........+.........................+.....................................................................................................................................................+.........+.......................................................................................................................................+.......................+............................................................................+.......................................................................+.......................+.......+............................................................................................................................................................................................................................................+....................................................................................................................+.....................................................................................................................................................................................................................................+..........................+............................+................................+...........................................................................................................................................................+...........................................................+........................+.........+........................................................................................................................................................................................................+......................................................................................................................................................................................................+.................................................................+.............................................+........................+..........................................................................................................................+.....+.........................................................+................................................+................................................................................+............................................................................................................................................................................................................................................+............................+........................................................+............................+...........+.........................+................................................+....................+........................................................................................................................+......................................+..............................................+...................................+...........................................................+........................................+.....+...................................+...............+..+...................................................................................................................................................................................+.............................+.....................................................................+.............................................................................................+...................................................................................................................................................................................................................+............................................+.....................................................................+..+......................+..................................................................................................................................+.....................................................................................................................................................................................................................................................................+.....+........................................................+..............................................................................................................................................................+......................................................................................+....................................................................................................................................................+.........................................................+......................................................................................................................................................................................................................................................................................................................................................................................................................+....+.............................................................................................................+.....................................................................................................................................................................+............................................................................................+...............................................................................................................................................................................+.................+............................................................................................................................................................................................................................+....................................................................................................................+..................................................................+...........................................................................................................................................................................................................+............................+...+........................................................................+..................................................................+..............................................+...........................................................................................+...................................................................................................................................+...............................................................................................................................................................................+........................................................................................................................................................+....+........................................................................+...................................................................................................................................................+..........................................................................................................................................................................................................................................................................................+....................................................................................................................................+..................................+....................................................................................................................................................................................................................................................................................................................................................................+..............................................................................................................................................+...............................................................................................+.......................................................................................................................................................................................+......+..................................................+...........................................................................................................................................................+................................................................................................................................................................................+..............................................................................................................................................+..................+..............................................................+.........................................................+......+.........+...............................................................................................+.........+.............................................................................................+................................................+......................................................................................................................+..........................................+................................................................................................................................+...........+..........................................................................................................+.........................+..............................................................................................................+................................................................................................................................................+........................+..................................+.........+..............................................................+..............+..........................................................................................................................................................+.......................................+........................................................+...................................................................................................................................................................................................................................+.................................................................................................+.........................................................................................................................................................................................................................................................+............................+..................................................+...............................................+..............................................................................................................................+...................................................................................................+......+........................................................................................................................................+.....................................................................+..............................................+..................................................................................................................................................................................................................+.............................................................................................................................+................................................................+...........................................................................................................................................+................................+.........................................................................................................................................................................................................................................................................................................................................................+...........+......................................+.........................................................................................+..............................................................................................................................................................................................................................................................................................................................................................................................................+....................................................................+......................................................................................+.......................................................................................................+................+.......+..+................................................................................+.........................................................+..........+.....+........................................+.........................................................................................................................................................................................................................................+...............................................................................................................................................................+.............................+................................................+...................................................................+..+.......+..............................................................+..................................................................+.....................+.........+............+.....................................................................................................+.........+........................................................................+.............+.....................................................................................+..................................................................+................................+..........................................................+...................................................................................................................................................................................................................+........................................................................................................................................+...........................................................................................................................................................+......................+..........................................................................+............................................................................+...............................................................................................................................................................................................................................................................................................+.....................................................................................................................................+.+........................................................+.........................+..............................................+......................+...........................................................+...............+...............................................................................+..................................................+.................................................................................+............................................................................................+....+...........................................................................................................+...+........................................................................................+...............................................+...+.......+...........................................................................+...................................................+.................................+...........................................+........................+..............................+..........................................................................................................................................................................................................................................................................................................................................................................................................+..+............+............................................+..................+...............................................................................................................+.......................................................................................................................................................................+.....................................................+...................................+......................................................................................................................................................................................+...................................+...............................................................................................................................................................................................+..+..............................................................................................+...+.........................................................................................+.....................................................................................................................+............................................+.......+..+.............................................................................................+..........................................................................................................................................................+..........+.............+.......................................................................................................................................................................................+....+.................................................................................+.........................................................................................................................................................................................+................................+..................+..................................................................+...............................+...............................................................+......................+..............+...................+.................................................................+..........................................................................................................................+.............................................................................................................................................................+......................................................................................................................+..............................................................+.......................................................................+........+..................+........................+...................................................+....................................................................................................+.+.......................................................................................................+.....................+............................................................+......+..................................................+..................................................+..........................................+...........................................................+.............................................................................................................................................................................................................................................................................................................................+..........................................................................+.......................................................................................................................+............................................................................................+................................................................................................................................................................................+.................................................................................+.......................................................................................................................................................................................................................................................................+......................................................................................................................................................................................................................................................................................................................................................+..................................+............................................................................................................................................................................................................................................................................................................................................+................................................................................................................................................................................................................+............................................................................................................................................+.....................................................................................................................................+....................................................................+.....................................+............+..............................................................................................................................................+...............................................................................+..................+.........................+........................+...................................+.....................................+...........+.................................+....................................................................................................................................................+.......................+......+....................................+........................................+..............................................................................................+........+......................................+...........................................................................................+...........................................+.........................................................................................................................................+............................................................................................................................................................................................................................................+.......................+......................................................................................+..............................................................................+.............+............................................................................................+..................................................................................+.................................................................................................................................+.......................................................................+....................................................................+...........................................+..................................................................................................................................................................................................................................................................................................................................................+.............................................................................+...........+..................................+.............................................................................................................................+.......................................................+..............................+..........................................................................................................+......................................................................................+..+....................................................................................................................................................................+.............................................................+...............................................................................................................................................................................+.................................................................................................................................................................................................+...................................................................................................................................................................................................................................................................................+..................................................................................................+.............................................................................+...........................+..................................................................................................................................................+............................+......+....................................................................................................................................................+.........+.................................................................................................................................................................................................................................................................................................................................................................................................................................................................+...........................................................................................................................................+............................................................................+..............+..............................................................................................+.....................+........................................................................................................................................................................+..........+.................................................................................................................................................................++*++*++*++*

DH parameters of size 2048 created at /etc/openvpn/pki/dh.pem

Note: using Easy-RSA configuration from: /etc/openvpn/vars
Using SSL: openssl OpenSSL 1.1.1g  21 Apr 2020
Generating a RSA private key
.................................................+++++
.............................................................+++++
writing new private key to '/etc/openvpn/pki/easy-rsa-68.GmDdej/tmp.cmbeMk'
-----
Using configuration from /etc/openvpn/pki/easy-rsa-68.GmDdej/tmp.KemKCF
Check that the request matches the signature
Signature ok
The Subject's Distinguished Name is as follows
commonName            :ASN.1 12:'VPN.SERVERNAME.COM'
Certificate is to be certified until Mar  6 07:36:25 2023 GMT (825 days)

Write out database with 1 new entries
Data Base Updated
Note: using Easy-RSA configuration from: /etc/openvpn/vars
Using SSL: openssl OpenSSL 1.1.1g  21 Apr 2020
Using configuration from /etc/openvpn/pki/easy-rsa-144.nagFbj/tmp.pCKmnb

An updated CRL has been created.
CRL file: /etc/openvpn/pki/crl.pem
```

**Comment by RafayAK** [comment]
Why is this issue closed? Just checked back, the problem is still there.

@kylemanna here are the images loaded on my computer:
![image](https://user-images.githubusercontent.com/17025191/100986561-2c8cd080-356f-11eb-9c14-e399538c7adf.png)

Following are the results of creating a client using different image tags:

- with `kylemanna/openvpn:latest`

![image](https://user-images.githubusercontent.com/17025191/100986902-9efdb080-356f-11eb-86e0-645bc687c945.png)

- with `kylemanna/openvpn:2.4`

![image](https://user-images.githubusercontent.com/17025191/100987073-d79d8a00-356f-11eb-98c7-7cce5f7be0b2.png)

- with `kylemanna/openvpn:2.3`

![image](https://user-images.githubusercontent.com/17025191/100987332-36630380-3570-11eb-8683-0be2f383b9a2.png)

As you can see only the image with the tag `2.3` passes successfully the rest fail.

**Comment by kylemanna** [comment]
@

**Comment by kylemanna** [comment]
@RafayAK thanks for sharing, now I understand what's going wrong.

Can you build the dev image and test it?
```
$ git fetch origin
$ git checkout origin/dev
$ docker build -t openvpn-test .
```

Then run the same command replacing `kylemanna/openpvn:2.4` with `openvpn-test` and share your result. Once I hear this works I can merge the dev branch and close this issue for good.

**Comment by RafayAK** [comment]
I'll try this tonight.

**Comment by RafayAK** [comment]
@kylemanna This seems to do the trick!

I created an image as you suggested:
![image](https://user-images.githubusercontent.com/17025191/101140254-87d4c680-3634-11eb-89cd-8b2e5ef42d9f.png)

Ran the command:
```bash
docker run -v $OVPN_DATA:/etc/openvpn --log-driver=none --rm -it openvpn-test easyrsa build-client-full $CLIENTNAME nopass
```
![image](https://user-images.githubusercontent.com/17025191/101140418-c1a5cd00-3634-11eb-8228-490766a34c45.png)

Now just to check if revocation and removal of a generated client key passes I also ran this command:
```bash
docker run --rm -it -v $OVPN_DATA:/etc/openvpn openvpn-test ovpn_revokeclient ${CLIENTNAME} remove
```

The output looks like this as follows:
![image](https://user-images.githubusercontent.com/17025191/101140841-514b7b80-3635-11eb-8e82-6d2cf16e6686.png)

This output is slightly different when compared to `kylemanna/openvpn:2.3`:
```bash
docker run --rm -it -v $OVPN_DATA:/etc/openvpn kylemanna/openvpn:2.3 ovpn_revokeclient ${CLIENTNAME} remove
```
![image](https://user-images.githubusercontent.com/17025191/101141141-af785e80-3635-11eb-9f8b-23160772086b.png)

the last 3 lines which show what files were removed is not showing, probably not a big deal though.
Either way thanks for fixing this.
